File: 21. Longest Consecutive Sequence/Solution.py
# set logic

# TC: O(3n)
# SC: O(n)


def lengthOfLongestConsecutiveSequence(nums, n) -> int:
    nums = set(nums)  # O(n)
    # print(nums)  # .... diff ... will see sorted +ve set then -ve set
    best = 0  # result ... longest streak

    # O(n) .... to visit each ele ... in loop
    # O(n) .... to check if next ele there .... create cycle
    for x in nums:
        # if x-1 in set .... means it had been traveled throw .... or will travel ... no need to visit it again
        # if x - 1 in nums:
        #     continue

        if x - 1 not in nums:
            y = x + 1  # next ele

            while y in nums:  # till next ele
                y += 1

            # max result store ... y-x ... diff between start and end .... is count
            # print(y - x) # count result
            best = max(best, y - x)

    return best

File: 21. Longest Consecutive Sequence/Solution.java
// sort logic

import java.util.Arrays;

public class Solution {
	public static int lengthOfLongestConsecutiveSequence(int[] nums, int n) {
		if (n == 0 || nums == null)
			return 0;

		Arrays.sort(nums);

		int longest = 1;
		int prev = nums[0];
		int cur_count = 1;

		for (int i = 1; i < n; i++) {
			if (nums[i] == prev + 1) {
				++cur_count;
			} else if (nums[i] != prev) {
				longest = Math.max(longest, cur_count);
				cur_count = 1;
			}

			prev = nums[i];
		}

		return Math.max(longest, cur_count);
	}
}
